pkg/controller/iam/managed.go

```diff
@@ -0,0 +1,232 @@
+func SetupServiceAccount(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1beta1.ServiceAccountGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1beta1.ServiceAccount{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1beta1.ServiceAccountGroupVersionKind),
+			managed.WithExternalConnecter(&connecter{client: mgr.GetClient(), newIAM: NewProviderClient}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connecter struct {
+	client client.Client
+	newIAM func(ctx context.Context, creds []byte) (ProviderClient, error)
+}
+
+// Connect sets up iam client using credentials from the provider
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1beta1.ServiceAccount)
+	if !ok {
+		return nil, errors.New(errNotServiceAccount)
+	}
+
+	p := &gcpv1alpha3.Provider{}
+	if err := c.client.Get(ctx, meta.NamespacedNameOf(cr.Spec.ProviderReference), p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errProviderSecretRef)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.client.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	iamAPI, err := c.newIAM(ctx, s.Data[p.Spec.CredentialsSecretRef.Key])
+	return &external{iam: iamAPI, projectID: p.Spec.ProjectID}, errors.Wrap(err, errNewClient)
+}
+
+type external struct {
+	iam       ProviderClient
+	projectID string
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) { // nolint:gocyclo
+	cr, ok := mg.(*v1beta1.ServiceAccount)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotServiceAccount)
+	}
+
+	fqName := FullyQualifiedName(*cr, e.projectID)
+	fromProvider, err := e.iam.GetServiceAccount(ctx, &adminpb.GetServiceAccountRequest{Name: fqName})
+	if status.Code(err) == codes.NotFound {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGet)
+	}
+
+	// TODO: write a test to assert that this gets updated in the reconciler
+	// from what i can tell every branch in the Reconcile code leads to a kube api update...
+	cr.Status.AtProvider = v1beta1.ServiceAccountObservation{
+		Name:           fromProvider.Name,
+		ProjectID:      fromProvider.ProjectId,
+		UniqueID:       fromProvider.UniqueId,
+		Email:          fromProvider.Email,
+		Oauth2ClientID: fromProvider.Oauth2ClientId,
+	}
+
+	rutd := isUpToDate(&cr.Spec.ForProvider, fromProvider)
+	cr.Spec.ForProvider.DisplayName = gcp.LateInitializeString(cr.Spec.ForProvider.DisplayName, fromProvider.DisplayName)
```

Comment:
Related to your `TODO` above, the `Status` gets updated in the reconciler code, but you will need to manually update the `Spec` here, which will also require passing a kube client to your `external` struct above

Reply:
I just took this logic out; considering the design patterns guidance that Spec belongs to crossplane, this should only be managed through crossplane. It's not one of those exceptional cases where something else in GCP could modify it as a side effect, so I'm saying not worth the cruft to check if it got modified out of band somehow.